extension/sigv4authextension/extension_test.go

```diff
@@ -0,0 +1,276 @@
+// limitations under the License.
+
+package sigv4authextension
+
+import (
+	"errors"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+
+	v4 "github.com/aws/aws-sdk-go/aws/signer/v4"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+)
+
+func TestNewSigv4Extension(t *testing.T) {
+	tests := []struct {
+		name          	string
+		cfg      	*Config
+		shouldError   	bool
+		expectedError 	string
+	}{
+		{
+			name: "valid_configurations",
+			cfg: &Config{
+				Region:     	"region",
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError:   false,
+			expectedError: "",
+		},
+		{
+			name: "missing_region",
+			cfg: &Config{
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError:   true,
+			expectedError: errNoRegionProvided.Error(),
+		},
+		{
+			name: "missing_service",
+			cfg: &Config{
+				Region:     	"region",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"regional",
+			},
+			shouldError:   true,
+			expectedError: errNoServiceProvided.Error(),
+		},
+		{
+			name: "invalid_endpoint",
+			cfg: &Config{
+				Region:     	"region",
+				Service: 	"service",
+				RoleArn:     	"rolearn",
+				STSEndpoint: 	"stsendpoint",
+			},
+			shouldError:   true,
+			expectedError: errInvalidSTSEndpoint.Error(),
+		},
+	}
+
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			sa, err := newSigv4Extension(testcase.cfg, "runtimeInfo", zap.NewNop())
+			if testcase.shouldError {
+				assert.NotNil(t, err)
+				assert.Contains(t, err.Error(), testcase.expectedError)
+				return
+			}
+			assert.NoError(t, err)
+			assert.Equal(t, testcase.cfg.Region, sa.cfg.Region)
+			assert.Equal(t, testcase.cfg.Service, sa.cfg.Service)
+			assert.Equal(t, testcase.cfg.RoleArn, sa.cfg.RoleArn)
+			assert.Equal(t, testcase.cfg.STSEndpoint, sa.cfg.STSEndpoint)
+		})
+	}
+}
+
+type ErrorRoundTripper struct{}
+
+func (ert *ErrorRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	return nil, errors.New("error")
+}
+
+func TestRoundTrip(t *testing.T) {
+	defaultRoundTripper := (http.RoundTripper)(http.DefaultTransport.(*http.Transport).Clone())
+	errorRoundTripper := &ErrorRoundTripper{}
+
+	tests := []struct {
+		name        	string
+		rt          	http.RoundTripper
+		shouldError 	bool
+		cfg  		*Config
+	}{
+		{
+			"valid_round_tripper",
+			defaultRoundTripper,
+			false,
+			&Config{Region: "region", Service: "service"},
+		},
+		{
+			"error_round_tripper",
+			errorRoundTripper,
+			true,
+			&Config{Region: "region", Service: "service", RoleArn: "rolearn"},
+		},
+	}
+
+	runtimeInfo := "runtimeInfo"
+
+	for _, testcase := range tests {
+		t.Run(testcase.name, func(t *testing.T) {
+			server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				_, err := v4.GetSignedRequestSignature(r)
```

Comment:
If you inject creds, you'll be able to verify the actual signature like here, not extremely helpful but may as well

https://github.com/aws/aws-sdk-go/blob/main/aws/signer/v4/v4_test.go#L112
https://github.com/aws/aws-sdk-go/blob/main/aws/signer/v4/v4_test.go#L202

Reply:
This was marked resolved but not addressed

Reply:
Took another look at this, and seems that the signature itself (`Signature="..."`) is always changing; I suspect it is either due to 1 of 2 things (or both):
1. `RoundTrip()` clones the request and signs that cloned request, or
2. in `Sign()`, we use `time.Now()`, while the tests you linked above use `time.Unix(0, 0)`; this seemed like the more likely of the two reasons, but upon testing it, the signature still changes every test. 
Not sure where to proceed from here; what do you think? If the signature is changing because of reason 1 (or both reasons), would it be valuable to make a separate struct/RoundTrip() implementation that doesn't clone the request (and possibly uses `time.Unix(0, 0)`), and test that instead?

Reply:
@Aneurysm9 looks like I'm injecting static credentials too, so theoretically that + using `time.Unix(0, 0)` should provide the same signature every time right?

Reply:
Ah thanks for clarifying the time.Time part, I think this is fine as is then